src/functions.php

```diff
@@ -285,6 +340,10 @@ function (mixed $throwable) use (&$rejected, &$rejectedThrowable, &$fiber): void
         throw $rejectedThrowable;
     }
 
+    if ($lowLevelFiber !== null) {
+        FiberMap::attachPromise($lowLevelFiber, $promise);
+    }
```

Comment:
Attaching for any promise and detaching for promises that are fulfilled seems to be implemented now, but looks like this is missing detaching for promises that are rejected? I'd like to see some additional tests that cover each case to ensure we do not keep any garbage references around.

Here's my basic idea, perhaps you can use this as a starting point for some additional tests?

```php
$fiber = null;
async(function () use (&$fiber) {
    $fiber = \Fiber::getCurrent();
    await(new Promise(function () { }));
})();

$this->assertCount(1, FiberMap::getPromises($fiber));
```

Reply:
@clue didn't add that initially because rejected is shortcircuiting the fiber anyway, but I forgot about try catch situations.  Just added a few tests based on your idea. However when disabling the detaching the tests still pass. So will have a look on the way home tonight. And maybe you have some ideas?

Reply:
Here's the basic idea again:

```php
    public function testAwaitRegistersPromiseInFiberMapUntilPromiseIsResolved()
    {
        $first = new Deferred();
        $second = new Deferred();
        $fiber = null;
        $promise = async(function () use (&$fiber, $await, $first, $second): int {
            $fiber = \Fiber::getCurrent();
            await($first->promise());
            await($second->promise());

            return time();
        })();

        $this->assertCount(1, React\Async\FiberMap::getPromises($fiber));
        $this->assertEquals([$first->promise()], array_values(React\Async\FiberMap::getPromises($fiber)));

        $first->resolve(null);
        $this->assertCount(1, React\Async\FiberMap::getPromises($fiber));
        $this->assertEquals([$second->promise()], array_values(React\Async\FiberMap::getPromises($fiber)));

        $second->resolve(null);
        $this->assertCount(0, React\Async\FiberMap::getPromises($fiber));
    }
```

The same idea applies to rejection, though it may involve slightly more logic to catch any exceptions. The above snippet also makes me wonder why we keep a list of promises for each fiber. It's my understanding each fiber can only wait for a single promise, no?

Reply:
Found out why tests keep passing no matter what.

Reply:
@clue Came to simpler tests. Commenting out the detaching will fail both tests:

```php
    /**
     * @dataProvider provideAwaiters
     */
    public function testResolvedPromisesShouldBeDetached(callable $await)
    {
        $await(async(function () use ($await): int {
            $fiber = \Fiber::getCurrent();
            $await(React\Promise\Timer\sleep(0.01));
            $this->assertCount(0, React\Async\FiberMap::getPromises($fiber));

            return time();
        })());
    }

    /**
     * @dataProvider provideAwaiters
     */
    public function testRejectedPromisesShouldBeDetached(callable $await)
    {
        $this->expectException(\Exception::class);
        $this->expectExceptionMessage('Boom!');

        $await(async(function () use ($await): int {
            $fiber = \Fiber::getCurrent();
            try {
                $await(React\Promise\reject(new \Exception('Boom!')));
            } catch (\Throwable $throwable) {
                throw $throwable;
            } finally {
                $this->assertCount(0, React\Async\FiberMap::getPromises($fiber));
            }

            return time();
        })());
    }
```

Reply:
> It's my understanding each fiber can only wait for a single promise, no?

Correct, working on that now :+1:

Reply:
> > It's my understanding each fiber can only wait for a single promise, no?
> 
> Correct, working on that now +1

@clue Done and pushed :point_up: